software/speller/data_collection_platform/dcp/bci/stream.py

```diff
@@ -21,51 +23,34 @@
 logger = logging.getLogger(__name__)
 logger.info('test logger')
 
-def stream_bci_api(bci_processes_states): 
-    # Some imports are only in this section, to allow other functions to be tested outside of the flask app
-    # The relative import paths make this difficult to test in isolation
-    # TODO: currently the collecting shared variable is in another branch, will have to test with other branch to see if this works
-    # from dcp.mp.shared import collecting
-    from dcp.mp.shared import q
-    # TODO: this next import has to do with the DB, and reading the BCI config ID
-    # will need to be done once the db is fully setup
-    # from dcp.models.configurations import OpenBCIConfig
-
-    # from multiprocessing import current_process
+def stream_bci_api(subprocess_dict): 
+
     logger.info('Attempting to initialize stream')
 
     inlet = connect_to_bci()
 
     logger.info('Succesfully connected to an inlet stream')
-    logger.info(inlet.info().as_xml())
-
-    # TODO: save current configuration to database once it is ready
-    # db.session.add(config)
-    # db.session.commit()
+    bci_config = inlet.info().as_xml()
+    logger.info(bci_config)
+    subprocess_dict['bci_config'] = bci_config
 
     # TAKEN FROM LAST YEAR:
     # retrieve estimated time correction offset for the given stream - this is
     # the number that needs to be added to a timestamp that was remotely
     # generated via local_clock() to map it into the local clock domain for
     # this machine
     inlet.time_correction()
+    subprocess_dict['state'] = 'ready'
+    queue = deque()
 
-    while bci_processes_states[os.getpid()] == False:
-        continue
-        # this will ensure that there is no race condition, and that the parent process has been able to instantiate this variable
-
-    # os.getpid()
-    # running = True
-    while bci_processes_states[os.getpid()]:
+    while subprocess_dict['state'] != 'stop':
         samples, _timestamps = inlet.pull_chunk()
-
-        if not samples:
+        if not samples or subprocess_dict['state'] != 'collect':
             continue
-        # TODO: uncomment this section to check the collecting value
-        # with collecting.get_lock():
-        #     if collecting.value:
-        q.put_nowait(samples)
-    
+
+        queue.append(samples)
+        subprocess_dict['q'] = queue
```

Comment:
@wykzhang why is this assignment within the while loop? any particular reason we want to assign the queue at every iteration? I'd recommend doing the assignment right after the queue was initialize, that way we always have a reference to this queue instance.

Reply:
from my understanding, the queue only gets passed as a proxy object between processes, and will therefore need to be constantly re-assigned for the parent process to have the right object. If we only assign the queue at the beginning, the parent process will only see an empty queue.

Reply:
![image](https://user-images.githubusercontent.com/43357040/149645668-4e1686de-6802-4b5f-a323-6ce40d0ea243.png)
I believe we don't have to reassign it again and again. If my experiment is correct, we should be able to directly append to the shared queue using subprocess_dict["q"].append(samples). Note I think `subprocess_dict` must be a `manager.dict()` object and `subprocess_dict["q"]` should be a `manager.Queue()` object.